ngraph/core/reference/include/ngraph/runtime/reference/deformable_convolution.hpp

```diff
@@ -0,0 +1,225 @@
+
+                inline Shape shape_scale(Shape s, size_t groups)
+                {
+                    s[0] /= groups;
+                    return s;
+                }
+
+                template <typename inputType>
+                inline float bilinear_interpolation(const inputType* data,
+                                                    const float x_idx,
+                                                    const float y_idx,
+                                                    const int x_size,
+                                                    const int y_size)
+                {
+                    const int x1 = std::max(static_cast<int>(std::floor(x_idx)), 0);
+                    const int x2 = std::min(static_cast<int>(std::ceil(x_idx)), x_size - 1);
+                    const int y1 = std::max(static_cast<int>(std::floor(y_idx)), 0);
+                    const int y2 = std::min(static_cast<int>(std::ceil(y_idx)), y_size - 1);
+
+                    const float distX = x_idx - x1;
+                    const float distY = y_idx - y1;
+
+                    const float value11 = data[y1 * x_size + x1];
+                    const float value12 = data[y2 * x_size + x1];
+                    const float value21 = data[y1 * x_size + x2];
+                    const float value22 = data[y2 * x_size + x2];
+
+                    const float value = (1 - distX) * (1 - distY) * value11 +
+                                        (1 - distX) * distY * value12 +
+                                        distX * (1 - distY) * value21 + distX * distY * value22;
+                    return value;
+                }
+
+                template <typename T>
+                void convolve_2D_channels(const ConvolutionParams& p,
+                                          const int64_t deformable_groups,
+                                          const T* batch,
+                                          const Shape& batch_shape,
+                                          const T* offsets,
+                                          const Shape& offset_shape,
+                                          const T* filter,
+                                          const Shape& filter_shape,
+                                          T* out)
+                {
+                    const int input_size_y = batch_shape[1];
+                    const int input_size_x = batch_shape[2];
+                    const int filter_size_y = filter_shape[1];
+                    const int filter_size_x = filter_shape[2];
+                    const int dilated_filter_size_y =
+                        filter_size_y + (filter_size_y - 1) * (p.dilation[0] - 1);
+                    const int dilated_filter_size_x =
+                        filter_size_x + (filter_size_x - 1) * (p.dilation[1] - 1);
+
+                    const int input_channel_size = shape_size(shape_reduce(batch_shape));
```

Comment:
As Shape contains `size_t` elements, is it okay to use simple `int` here and below?

Reply:
I deliberately chosen to convert every variable to int to avoid issues with implicit conversions and make it easier for the reader at the same time (he/she doesn't need to be concerned if types matter in those calculations). [Here](https://github.com/openvinotoolkit/openvino/blob/master/ngraph/core/reference/include/ngraph/runtime/reference/convolution.hpp#L255) is the comment with explanation, the same approach is in DeformableConvolution. [Here](https://stackoverflow.com/questions/17832815/c-implicit-conversion-signed-unsigned) is some discussion on the topic. In general I don't see any functional issue with it and if you do please reopen and point me exactly what's wrong with this approach.